packages/emotion-theming/types/index.d.ts

```diff
@@ -1,34 +1,29 @@
-// TypeScript Version: 2.3
-import { ComponentClass, SFC } from "react";
+// Definitions by: Junyoung Clare Jang <https://github.com/Ailrun>
+// TypeScript Version: 2.8
 
-export type OptionalThemeProps<Props, Theme> = Props & { theme?: Theme };
+import * as React from 'react'
+
+import { AddOptionalTo, PropsOf } from './helper'
 
 export interface ThemeProviderProps<Theme> {
-    theme: Partial<Theme> | ((theme: Theme) => Theme);
+  theme: Partial<Theme> | ((outerTheme: Theme) => Theme)
 }
 
-export type ThemeProviderComponent<Theme> = ComponentClass<ThemeProviderProps<Theme>>;
-export const ThemeProvider: ThemeProviderComponent<object>;
+export function ThemeProvider<Theme>(
+  props: ThemeProviderProps<Theme>
+): React.ReactElement<any>
```

Comment:
Why would someone want to provide a type to this? how would it actually give them any type safety? the current theming API is inherently not type safe so I'm not a huge fan of trying to add a bunch of types to it if it doesn't actually give any type safety

Reply:
It provides minimal type safety, namely, when one already has some `Theme` type,
`ThemeProvider<Theme>` will reject following cases. (Let's say `Subtype` as a subtype of `Theme`, and supertype as a supertype of `Theme`.)
```tsx
<ThemeProvider<Theme> theme={(outerTheme: Subtype): Supertype => { /* ... */ }}>
  {/* ... */}
</ThemeProvider>
```
```tsx
const superTheme: Supertype = /* ... */
<ThemeProvider<Theme> theme={superTheme}>
  {/* ... */}
</ThemeProvider>
```